keps/sig-api-machinery/4595-cel-crd-additionalprintercolumns/README.md

```diff
@@ -0,0 +1,1447 @@
+
+```go
+func (c ColumnCompilationResult) FindResults(data interface{}) ([][]reflect.Value, error) {
+  ...
+}
+
+func (c ColumnCompilationResult) PrintResults(w io.Writer, results []reflect.Value) error {
+  ...
+}
+```
+
+- The output of `cel.CompileColumn()` returns a `ColumnCompilationResult` object for each additionalPrinterColumn. 
+
+With all of this we can pass the CEL program to the TableConvertor's `ConvertToTable()` method, which will call `FindResults` and `PrintResults` for all additionalPrinterColumns, regardless of whether they're defined with JSONPath or CEL expressions.
+
+
+### CEL vs JSONPath Performance Analysis
+
+A big part of the discussions for our proposal was the CEL cost limits since this is the first time CEL is added to the read path. As part of this we've done some benchmarking of the time it takes to parse and compile equivalent JSONPath and CEL expressions.
+
+For the performance analysis, we introduced two new benchmark tests to the `tableconvertor_test.go` file – `Benchmark_CEL` and `Benchmark_JSONPath`.
+
+Please find the raw output of the benchmark tests, as well as the code for executing them in the following gist:
+[https://gist.github.com/sreeram-venkitesh/f4aff1ae7957a5a3b9c6c53e869b7403](https://gist.github.com/sreeram-venkitesh/f4aff1ae7957a5a3b9c6c53e869b7403)
+
+
+Considering the following 10 iterations of the `Benchmark_CEL` and `Benchmark_JSONPath` tests:
+
+```
+  CEL Benchmark Results (10 runs):
+  - Run 1: 2,772 iterations, 461,257 ns/op
+  - Run 2: 3,070 iterations, 355,537 ns/op
+  - Run 3: 3,247 iterations, 353,626 ns/op
+  - Run 4: 3,090 iterations, 360,392 ns/op
+  - Run 5: 3,249 iterations, 351,305 ns/op
+  - Run 6: 2,942 iterations, 454,657 ns/op
+  - Run 7: 3,379 iterations, 354,431 ns/op
+  - Run 8: 2,858 iterations, 365,886 ns/op
+  - Run 9: 3,374 iterations, 353,106 ns/op
+  - Run 10: 3,130 iterations, 368,940 ns/op
+
+  JSONPath Benchmark Results (10 runs):
+  - Run 1: 76,616 iterations, 19,376 ns/op
+  - Run 2: 73,274 iterations, 15,761 ns/op
+  - Run 3: 72,625 iterations, 18,572 ns/op
+  - Run 4: 70,324 iterations, 15,896 ns/op
+  - Run 5: 69,409 iterations, 15,892 ns/op
+  - Run 6: 70,672 iterations, 17,888 ns/op
+  - Run 7: 72,145 iterations, 16,176 ns/op
+  - Run 8: 69,602 iterations, 16,203 ns/op
+  - Run 9: 68,079 iterations, 24,558 ns/op
+  - Run 10: 62,689 iterations, 16,214 ns/op
+```
+
+The following table provides an average performance analysis across CEL and JSONPath based additionalPrinterColumns:
+
+
+|                                      | CEL (Benchmark_CEL)                                                                                                   | JSONPath (Benchmark_JSONPath)                                                                                   |
```

Comment:
Instead of post-processing the data, it might be better to have separate benchmarks for compile versus runtime.

Also, what hardware / architecture was this run on?

Reply:
> Instead of post-processing the data, it might be better to have separate benchmarks for compile versus runtime.

Ack. We are working on benchmarking more expensive use cases per above thread's discussion.
Will address adding separate compilation and runtime benchmark stats as part of the same update.


> Also, what hardware / architecture was this run on?

The current benchmarks stats are from an arm64 machine (Apple M3 Pro with 12 cores, 18 GB RAM).

We also have access to an x86_64 machine (11th Gen Intel Core i7-11800H × 16 cores, 48 GB RAM) if further stats needed.